### ISSUE TITLE
We should relax the requirement that EntryPointAttribute only exists in .exe files

### ISSUE THREAD

**Issue by KevinRansom** [Microsoft/visualfsharp#1150]
coreclr exe's can be .exe files or .dll files.  renaming corehost.exe myprog.exe, will cause it to search for a dll named myprog.dll and execute that.

I have been using this behavior for a long time on coreclr builds, by renaming the output files after building. them.

WEell that seems kind of hacky, so I am going to modify the build to emit dll's eliminating the "Hack"

In order to this, I need to relax this error:

error FS0433: A function labeled with the 'EntryPointAttribute' attribute must be the last declaration in the last file in the compilation sequence**, and can only be used when compiling to a .exe **

to eliminate the bold text and the test for it.  C# silently allows a .dll to contain a main method, and so I see no good reason for maintaining this behavior.

**Comment by dsyme** [comment]
No objections at all.  Though can/should we make it dependent on compiling to CoreCLR?  The reason to maintain it would be that people accidentally think they are making a .EXE when they aren't, and they get no warning or error (I know C# doesn't do that, but C# isn't perfect, and doesn't have an explicit EntryPoint attribute).

**Comment by KevinRansom** [comment]
C# has an explicit Main, which is the moral equivalent of F#'s entrypoint attribute.  I will always relax it because the compiler won't know it's targeting coreclr.

**Comment by KevinRansom** [comment]
I will make this a warning :
````
A function labeled with the 'EntryPointAttribute' attribute must be the last declaration in the last file in the compilation sequence, and can only be used when compiling to a .exe
````
as well as this: 
````
'Files in libraries or multiple-file applications must begin with a namespace or module declaration, e.g. 'namespace SomeNamespace.SubNamespace' or 'module SomeNamespace.SomeModule'. Only the last source file of an application may omit such a declaration.'
````

**Comment by dsyme** [comment]
Is there any reason not to drive this based on the explicit use of ``--target:exe`` ?

Making these warnings doesn't feel right.  Either you're allowed to do these things in compilations or not.   (For the second, we could live with it being a warning for the last file in a DLL, but not earlier files).

BTW I thought the compiler knows it is targeting CoreCLR through ``--targetprofile:netcore``?

> C# has an explicit Main, which is the moral equivalent of F#'s entrypoint attribute.

C# DLLs can (and do) have methods called Main in ordinary classes.  F#'s ``EntryPoint`` is much more explicit, and it's a clear error to use it when it's not the last declaration (and in DLLs when not targeting .NET Core or not using ``--target:exe``).

**Comment by enricosada** [comment]
@dsyme .net cli pass `--target:exe` too if it's an executable (instead of library) for entrypoint creation.

**Comment by KevinRansom** [comment]
@dsyme I think a warning is for when you are doing something that you should think about.  like the old precision losing casts from double to float to int etc....

In the case of the entry point not being at the end of the last file.  Well ... you probably want to know about it in case it was unintended.  But there may be circumstances where it was intended.   For example the program was intended to be used as either an application or as a library and the interop APIs appeared after the Main function, because they were never intended to be used from within the app. 

In this case I just want to be able to build an application into a .dll during the build.  The compiler rather unhelpfully says I can't do that.

I can't think of any C# source code that will compile when targeting an exe and not compile when targeting a dll so I think that the behavior should at least be consistent regardless of target.

--target:netcore tells the compiler to use the portable libraries rather than mscorlib.  It is not coreclr specific.   In my fondest dreams, we will get rid of all of those hacks and reference types cleanly eventually.  Although for now we are stuck with it.

I can try to keep the errors as errors but will try to lose the limitation of complaining about the .dll.  I gave up on figuring that out a bit too quickly yesterday.

**Comment by KevinRansom** [comment]
Groan !!!!
it was easy when I thought about it right.

**Comment by dsyme** [comment]
@KevinRansom What was the resolution?  Can we close this?

**Comment by enricosada** [comment]
ref #824 
fixed by #1159

i think can be closed @dsyme
